src/shared_types.rs

```diff
@@ -56,6 +58,51 @@ pub struct Color {
     pub alpha: f32,
 }
 
+impl FromStr for Color {
+    type Err = ErrorKind;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let colors: Vec<f32> = s.split(',').map(|v| v.parse().unwrap()).collect();
```

Comment:
I think it's worth skipping this allocation? I'm not sure if you can `collect` to a fixed-size array, but that might be worth it, or you could work with the iterator directly. Here's a quick sketch of how I might write this:

```rust
let mut iter = s.split(',').map(|v| match v.parse::<f32>() {
    Ok(val) if (0.0..=1.0).contains(val) => Ok(val),
    _ => ErrorKind::BadColor,
});
let red = iter.next()?;
let green = iter.next()?;
let blue = iter.next()?;
let alpha = iter.next()?;
if iter.next().is_some() {
    Err(ErrorKind::BadColor)
} else {
    Ok(Color { red, green, blue, alpha })
}
```